File: leet_code/1161. Maximum Level Sum of a Binary Tree.py

import time
from util.util_list import *
from util.util_tree import *
import copy
import heapq
import bisect
import collections


class Solution:
    def maxLevelSum(self, root: TreeNode) -> int:
        q = [(1, root)]
        levels = collections.defaultdict(int)
        
        while q:
            nq = []
            
            while q:
                lev, node = q.pop(0)
                levels[lev] += node.val
                
                if node.left:
                    nq += (lev + 1, node.left),
                
                if node.right:
                    nq += (lev + 1, node.right),
            
            q = nq
        
        levels = sorted(levels.items(), key=lambda p: p[1], reverse=True)
        return levels[0][0]
        

stime = time.time()
print(2 == Solution().maxLevelSum(deserialize('[1,7,0,7,-8,null,null]')))
print('elapse time: {} sec'.format(time.time() - stime))

File: leet_code/1161. Maximum Level Sum of a Binary Tree.js


var maxLevelSum = function(root) {
    let q = [[1, root]]
    levels = {}

    while (q.length > 0) {
        let nq = []

        while (q.length > 0) {
            item = q.shift()
            let lev = item[0]
            let node = item[1]

            if (!(lev in levels))
                levels[lev] = 0

            levels[lev] += node.val

            if (node.left)
                nq.push([lev + 1, node.left])

            if (node.right)
                nq.push([lev + 1, node.right])
        }

        q = nq
    }
    
    var kvs = []
    for (var key in levels)
        kvs.push([key, levels[key]])

    kvs.sort(function compare(l, r) {
        return l[1] - r[1]
    });

    return kvs[kvs.length - 1][0]
}


